skimage/transform/_thin_plate_splines.py

```diff
@@ -0,0 +1,316 @@
+    ----------
+    r : (4, N) ndarray
+        Input array representing the norm distance between interlandmark
+        distances for the source form and based on the (x,y) coordinates
+        for each of these points.
+
+    Returns
+    -------
+    U : (4, N) ndarray
+        Calculated kernel function U.
+    """
+    _small = 1e-8  # Small value to avoid divide-by-zero
+    r_sq = r**2
+    U = np.where(r == 0.0, 0.0, r_sq * np.log(r_sq + _small))
+    return U
+
+
+def tps_warp(
+    image, src, dst, *, output_region=None, interpolation_order=1, grid_scaling=1
+):
+    """Return an array of warped images.
+
+    Define a thin-plate-spline warping transform that warps from the
+    src to the dst, and then warp the given images by
+    that transform.
+
+    Parameters
+    ----------
+    image : ndarray
+        Input image.
+    src : (N, 2)
+        Control points at source coordinates.
+    dst : (N, 2)
+        Control points at target coordinates.
+    output_region : tuple of integers, optional
+        The region ``(xmin, ymin, xmax, ymax)`` of the output
+        image that should be produced. (Note: The region is inclusive, i.e.
+        xmin <= x <= xmax)
+    interpolation_order : int, optional
+        If 1, use linear interpolation, otherwise use
+        nearest-neighbor interpolation.
+    grid_scaling : int, optional
+        If grid_scaling is greater than 1, say x, then the transform is
+        defined on a grid x times smaller than the output image region.
+        Then the transform is bilinearly interpolated to the larger region.
+        This is fairly accurate for values up to 10 or so. Must be equal to or
+        greater than 1.
+
+    Returns
+    -------
+    warped : array_like
+        The warped input image.
+
+
+    References
+    ----------
+    .. [1] Bookstein, Fred L. "Principal warps: Thin-plate splines and the
+           decomposition of deformations," IEEE Transactions on pattern analysis
+           and machine intelligence 11.6 (1989): 567–585.
+
+    Examples
+    --------
+    Produce a warped image rotated by 90 degrees counter-clockwise.
+
+    >>> import skimage as ski
+    >>> astronaut = ski.data.astronaut()
+    >>> image = ski.color.rgb2gray(astronaut)
+    >>> src = np.array([[0, 0], [0, 512], [512, 512],[512, 0]])
+    >>> dst = np.array([[512, 0], [0, 0], [0, 512],[512, 512]])
+    >>> output_region = (0, 0, image.shape[0], image.shape[1])
+    >>> warped_image = ski.transform.tps_warp(
+    ...     image, src, dst, output_region=output_region
+    ... )
+
+    """
+    image = np.asarray(image)
+
+    # Check if grid_scaling is equal to or greater than 1
+    if grid_scaling < 1:
+        raise ValueError("Grid scaling must be equal to or greater than 1.")
+
+    if image.size == 0:
+        raise ValueError(f"Cannot warp empty image with dimensions {image.shape!r}")
+
+    if image.ndim not in (2, 3):
+        raise ValueError("Only 2D and 3D images are supported")
```

Comment:
Oh, I got the impression that only 2D images were supported when reading about (x, y) coordinates...